streams/src/test/java/org/apache/kafka/streams/integration/StoreUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,815 @@
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Properties;
+
+@Category({IntegrationTest.class})
+public class StoreUpgradeIntegrationTest {
+    private static String inputStream;
+    private static final String STORE_NAME = "store";
+
+    private KafkaStreams kafkaStreams;
+    private static int testCounter = 0;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(1);
+
+    @Before
+    public void createTopics() throws Exception {
+        inputStream = "input-stream-" + testCounter;
+        CLUSTER.createTopic(inputStream);
+    }
+
+    private Properties props() {
+        final Properties streamsConfiguration = new Properties();
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, "addId-" + testCounter++);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.STATE_DIR_CONFIG, TestUtils.tempDirectory().getPath());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 1000);
+        streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        streamsConfiguration.put(IntegrationTestUtils.INTERNAL_LEAVE_GROUP_ON_CLOSE, true);
+        return streamsConfiguration;
+    }
+
+    @After
+    public void shutdown() {
+        if (kafkaStreams != null) {
+            kafkaStreams.close(Duration.ofSeconds(30L));
+            kafkaStreams.cleanUp();
+        }
+    }
+
+    @Test
+    public void shouldMigrateKeyValueStoreToTimestampedKeyValueStoreUsingPapi() throws Exception {
+        final StreamsBuilder streamsBuilderForOldStore = new StreamsBuilder();
+
+        streamsBuilderForOldStore.addStateStore(Stores.keyValueStoreBuilder(
+            Stores.persistentKeyValueStore(STORE_NAME),
+            Serdes.Integer(),
+            Serdes.Long()
+        ));
+        streamsBuilderForOldStore
+            .<Integer, Integer>stream(inputStream)
+            .process(KeyValueProcessor::new, STORE_NAME);
+
+        final Properties props = props();
+        kafkaStreams = new KafkaStreams(streamsBuilderForOldStore.build(), props);
+        kafkaStreams.start();
+
+        final LinkedList<KeyValue<Integer, Object>> expectedResult = new LinkedList<>();
+
+        expectedResult.add(KeyValue.pair(1, 1L));
+        processKeyValueAndVerifyPlainCount(1, expectedResult);
+
+        expectedResult.remove(0);
```

Comment:
nit: I'm having a hard time maintaining a clear mental picture of what the expected result looks like as this test progresses. Can you consider using a literal list for each `processKeyValueAndVerifyPlainCount` instead?